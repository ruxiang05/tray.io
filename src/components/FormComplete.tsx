import React from "react";
import { ReactComponent as CheckIcon } from "../assets/check-solid.svg";

const Done = () => (
  <div className="form-complete">
    <CheckIcon className="check-icon" />
    <p>
      Please verify your email address, you should have received an email from
      us already!
    </p>
  </div>
);

export default Done;
