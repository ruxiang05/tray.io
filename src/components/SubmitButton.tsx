import React from "react";

interface SubmitButtonProps {
  changePage: (a: string) => void;
  disabled?: boolean;
}

const SubmitButton = ({ changePage, disabled }: SubmitButtonProps) => (
  <button
    type="button"
    onClick={() => changePage("privacy")}
    disabled={disabled}
  >
    Submit
  </button>
);

export default SubmitButton;
