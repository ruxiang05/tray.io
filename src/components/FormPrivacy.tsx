import React from "react";
import FormInput from "./FormInput";
import { FormProps } from "./Form";
import SubmitButton from "./SubmitButton";

const Privacy = ({ changePage, handleChange }: FormProps) => (
  <div>
    <FormInput
      type="checkbox"
      name="productComm"
      text="Receive updates about Tray.io product by email"
      handleChange={handleChange}
    />
    <FormInput
      type="checkbox"
      name="otherComm"
      text="Receive communication by email for other products created by the Tray.io
      team"
      handleChange={handleChange}
    />
    <SubmitButton changePage={() => changePage("done")} />
  </div>
);

export default Privacy;
