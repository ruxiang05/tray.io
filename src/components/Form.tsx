import React, { useState } from "react";
import FormUser from "./FormUser";
import FormHeader from "./FormHeader";
import FormPrivacy from "./FormPrivacy";
import FormComplete from "./FormComplete";
import "../styles/styles.scss";

export interface FormProps {
  error?: string;
  changePage: (a: string) => void;
  handleChange: (e: React.FormEvent<HTMLInputElement>) => void;
}

const Form = () => {
  const [formUserState, setFormUserState] = useState({
    name: "",
    role: "",
    email: "",
    password: "",
  });
  const [page, setPage] = useState("user");
  const [error, setError] = useState("");

  const handleChange = (e: React.FormEvent<HTMLInputElement>) => {
    const { value, name } = e.currentTarget;

    setFormUserState({
      ...formUserState,
      [name]: value,
    });
  };

  const changePage = (page: string) => {
    const isValid = validateForm();

    if (isValid) setPage(page);
  };

  const validateForm = () => {
    if (!(validateName() && validateEmail() && validatePassword())) {
      setError("Form is incorrect, please check again");
      return false;
    }

    return true;
  };

  const validateName = (): boolean => {
    const { name } = formUserState;

    return name ? true : false;
  };

  const validateEmail = (): boolean => {
    const { email } = formUserState;
    if (!email) {
      return false;
    }

    const regex = new RegExp(/\S+@\S+\.\S+/);
    return regex.test(email);
  };

  const validatePassword = (): boolean => {
    const { password } = formUserState;

    if (!password) {
      return false;
    }

    const regex: RegExp = new RegExp(/^(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{9,}$/g);
    return regex.test(password);
  };

  const renderFormPage = (page: string) => {
    switch (page) {
      case "user":
        return (
          <FormUser
            changePage={changePage}
            handleChange={handleChange}
            error={error}
          />
        );
      case "privacy":
        return (
          <FormPrivacy changePage={changePage} handleChange={handleChange} />
        );
      case "done":
        return <FormComplete />;
    }
  };

  return (
    <div className="form">
      <FormHeader currentPage={page} />
      <form>{renderFormPage(page)}</form>
    </div>
  );
};

export default Form;
