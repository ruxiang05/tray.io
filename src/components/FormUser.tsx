import React from "react";
import FormInput from "./FormInput";
import SubmitButton from "./SubmitButton";
import { FormProps } from "./Form";

const FormUser = ({ changePage, handleChange, error }: FormProps) => (
  <div>
    <FormInput
      type="text"
      name="name"
      placeholder="Name"
      required
      handleChange={handleChange}
    />
    <FormInput
      type="text"
      name="role"
      placeholder="Role"
      handleChange={handleChange}
    />
    <FormInput
      type="email"
      name="email"
      placeholder="Email Address"
      required
      handleChange={handleChange}
    />
    <FormInput
      type="password"
      name="password"
      placeholder="Password"
      required
      handleChange={handleChange}
    />
    {error && <p className="form-error">{error}</p>}
    <SubmitButton changePage={changePage} disabled={error ? true : false} />
  </div>
);

export default FormUser;
