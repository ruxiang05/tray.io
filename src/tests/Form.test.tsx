import React from "react";
import renderer from "react-test-renderer";
import { shallow, mount } from "enzyme";
import Form from "../components/Form";
import FormUser from "../components/FormUser";
import FormPrivacy from "../components/FormPrivacy";
import FormComplete from "../components/FormComplete";
import SubmitButton from "../components/SubmitButton";

describe("Form", () => {
  const mock = {
    changePage: jest.fn(),
    handleChange: jest.fn(),
    error: "undefined",
  };

  it("should render", () => {
    const tree = renderer.create(<Form />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render the correct form page", () => {
    const wrapper = mount(<Form />);
    let submitButton = wrapper.find(SubmitButton);
    expect(submitButton).toBeDefined();
    submitButton.simulate("click");
    expect(wrapper.find(FormPrivacy)).toBeDefined();
  });
});
