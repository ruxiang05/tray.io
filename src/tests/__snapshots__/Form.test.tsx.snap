// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render 1`] = `
<div
  className="form"
>
  <ul
    className="form-header"
  >
    <li
      className="selected-page"
    >
      user
    </li>
    <li
      className=""
    >
      privacy
    </li>
    <li
      className=""
    >
      done
    </li>
  </ul>
  <form>
    <div>
      <div
        className="form-input"
      >
        <label
          htmlFor="name"
        >
          name:
        </label>
        <span
          className="required-mark"
        >
           * 
        </span>
        <input
          name="name"
          onChange={[Function]}
          placeholder="Name"
          required={true}
          type="text"
        />
      </div>
      <div
        className="form-input"
      >
        <label
          htmlFor="role"
        >
          role:
        </label>
        <input
          name="role"
          onChange={[Function]}
          placeholder="Role"
          type="text"
        />
      </div>
      <div
        className="form-input"
      >
        <label
          htmlFor="email"
        >
          email:
        </label>
        <span
          className="required-mark"
        >
           * 
        </span>
        <input
          name="email"
          onChange={[Function]}
          placeholder="Email Address"
          required={true}
          type="email"
        />
      </div>
      <div
        className="form-input"
      >
        <label
          htmlFor="password"
        >
          password:
        </label>
        <span
          className="required-mark"
        >
           * 
        </span>
        <input
          name="password"
          onChange={[Function]}
          placeholder="Password"
          required={true}
          type="password"
        />
      </div>
      
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
